plugin/src/main/java/org/opensearch/ml/action/stats/MLStatsNodesTransportAction.java

```diff
@@ -126,22 +146,72 @@ MLStatsNodeResponse createMLStatsNodeResponse(MLStatsNodesRequest mlStatsNodesRe
             }
         }
 
+        CountDownLatch latch = new CountDownLatch(mlStats.getAllModels().length);
         Map<String, MLModelStats> modelStats = new HashMap<>();
         // return model level stats
         if (mlStatsInput.includeModelStats()) {
             for (String modelId : mlStats.getAllModels()) {
-                if (mlStatsInput.retrieveStatsForModel(modelId)) {
-                    Map<ActionName, MLActionStats> actionStatsMap = new HashMap<>();
-                    for (Map.Entry<ActionName, MLActionStats> entry : mlStats.getModelStats(modelId).entrySet()) {
-                        if (mlStatsInput.retrieveStatsForAction(entry.getKey())) {
-                            actionStatsMap.put(entry.getKey(), entry.getValue());
+                // Add action listener to retrieve model details
+                validateAccess(modelId, ActionListener.wrap(hasPermissionToShowStat -> {
+                    if (hasPermissionToShowStat) {
+                        if (mlStatsInput.retrieveStatsForModel(modelId)) {
+                            Map<ActionName, MLActionStats> actionStatsMap = new HashMap<>();
+                            for (Map.Entry<ActionName, MLActionStats> entry : mlStats.getModelStats(modelId).entrySet()) {
+                                if (mlStatsInput.retrieveStatsForAction(entry.getKey())) {
+                                    actionStatsMap.put(entry.getKey(), entry.getValue());
+                                }
+                            }
+                            modelStats.put(modelId, new MLModelStats(actionStatsMap));
                         }
                     }
-                    modelStats.put(modelId, new MLModelStats(actionStatsMap));
-                }
+                    // Count down the latch after each asynchronous call completes
+                    latch.countDown();
+                }, e -> {
+                    // Handle failure case here
+                    // For example, log the error
+                    log.error("Failed to retrieve model details for model ID: " + modelId, e);
+                    // Count down the latch even in case of failure to ensure proper synchronization
+                    latch.countDown();
+                }));
+            }
+            // Wait for all asynchronous calls to complete
+            try {
+                latch.await();
+            } catch (InterruptedException e) {
+                // Handle interruption if necessary
+                Thread.currentThread().interrupt();
             }
         }
-
         return new MLStatsNodeResponse(clusterService.localNode(), statValues, algorithmStats, modelStats);
     }
+
+    private void validateAccess(String modelId, ActionListener<Boolean> listener) {
+        boolean isSuperAdmin = isSuperAdminUserWrapper(clusterService, client);
+        String[] excludes = new String[] { MLModel.MODEL_CONTENT_FIELD, MLModel.OLD_MODEL_CONTENT_FIELD };
```

Comment:
Looks like only the isHidden flag is needed. So why not only return this flag in the MLModel by defining an String[] include?

Reply:
the underlying method is checking algorithm field. So I don't think adding includes will be a good option here. Do you see any risk of adding `excludes`? I don't think it will add any extra latency in the query.